### PR TITLE
[KAN-4] Changed get_notams to request one location's notams.

### PR DESCRIPTION
**Pull request by Zombiepro89**
This is useful so that the function is generic,
thus it can be used for depature, arrival, and in-flight notam requests.

Added a skeleton function for front-end and back-end connection. Once we're ready to connect the front and back-end, we can call the start_notam_request_path() function which will call the necessary functions and return the output the front-end is expecting.
This function should be built on in future commits.

### CODE REVIEW

File: notamFetch.py [hunk 1]
```diff
@@ -3,60 +3,122 @@
...
+    return notam_list
+
+"""This is the starting point for the program, the front end should call this function.
+    From there, this function should call other functions to 
+    retrieve depature and arrival airport notams as well as in-flight notams,
+    get these notams sorted, and return the sorted list back to the front end."""
+
+    # TODO: Currently returns the union of the depature and arrival airport notams.
+    # Looking to add in-flight notams and figure out a way to remove any intersecting notams.
+def start_notam_request_path(departure_airport, arrival_airport) -> list:
```
Comment (JamesJWilliams3000) [inline notamFetch.py:116]: The function name seems unclear to me, if this is the function meant to be called by the frontend. `start_notam_request_path` doesn't imply that it is going to return a sorted list of notams.

File: notamFetch.py [hunk 2]
```diff
@@ -3,60 +3,122 @@
...
+# this can be called anywhere doesnt have to be in the function call here
+# credentials retrieved from external '.env' file located in root dir
+credentials = {
+    "client_id": os.getenv('client_id'),
+    "client_secret": os.getenv('client_secret'),
+}
+
+""" From a given airport code or longitude and latitude tuple, return a list of notams from the api.
+    airport_code expects a string of some kind.
+    request_longitude_latitude expects a tuple of the form (log : float, lat : float)
```
Comment (JamesJWilliams3000) [inline notamFetch.py:25]: I feel like `request_longitude_latitude` needs to be an explicit object, both for clearer documentation and to prevent errors. For example:
`request_longitude_latitude.latitude` and `request_longitude_latitude.longitude` are clearer than `request_longitude_latitude[0]` and `request_longitude_latitude[1]`. This also prevents a longitude being assigned to `request_longitude_latitude[0]` by mistake.

As an object, we can also constrain `request_longitude_latitude` values to float, so that we get an error when the object is initialized, as opposed to error checking anytime we need to use `request_longitude_latitude`.

File: notamFetch.py [hunk 3]
```diff
@@ -3,60 +3,122 @@
...
+    """
+def get_notams(airport_code = None, request_longitude_latitude = None) -> list:
+    
+    # Check to see if the user inputted an airport code, 
+    # if so request geolocator to find the longitude and latitude.
+    if(not (airport_code is None)):
```
Comment (JamesJWilliams3000) [inline notamFetch.py:31]: This isn't easy to read. (genuinely trying to be helpful, not mean)

File: notamFetch.py [hunk 4]
```diff
@@ -3,60 +3,122 @@
...
+            print("Invalid airport.", airport_code, "was not found.")
+            return None
+            
+        
+    # If the user did not enter an airport code or geolocator successfully outputted something,
+    # check to make sure the request is in the correct form
+    if(type(request_longitude_latitude) is tuple and len(request_longitude_latitude) == 2 
+            and (type(request_longitude_latitude[0]) is float and type(request_longitude_latitude[1]) is float)):
+        
+        return send_api_request(request_longitude_latitude)
```
Comment (JamesJWilliams3000) [inline notamFetch.py:47]: I really think we should use method overloading here and try to minimize the number of things `get_notams()` is doing.
My idea:
```
def get_notams(request_longitude_latitude) -> list:
    # error handling, etc
    return send_api_request(request_longitude_latitude)
```
```
def get_notams(airport_code) -> list:
    # get `request_longitude_latitude` from `airport_code`, or raise an Exception if invalid
    return get_notams(request_longitude_latitude)
```

^ note at the end I called `get_notams` and not `send_api_request`, so that we guarantee the same functionality across both `get_notams`, if we ever make changes in the future.

Comment (Zombiepro89) [reply notamFetch.py:47]: That was the idea at first, but, unfortunately, python does not support overloading functions, so this is the best next option. I don't like that I have to have that if statement at the beginning, but, without extra libraries, this seems to be the best way to do it. But I do absolutely agree with your ideas.

**Review by JamesJWilliams3000 — CHANGES_REQUESTED** [review]
I left a few comments for my review, mostly regarding `request_longitude_latitude`. 

I also think we should reconsider our function names, as `start_notam_request_path` and `get_notams` don't seem very clear on what they do at the moment. It doesn't seem obvious that `get_notams` shouldn't be called from the frontend, and `start_notam_request_path` doesn't sound like a function that will return a complete and sorted list of notams to the frontend.

File: notamFetch.py (hunk 3 above)
Comment (Zombiepro89) [reply notamFetch.py:31]: Is it the parentheses that make it difficult to read? Would it be better as `not airport_code is None`?

File: notamFetch.py (hunk 2 above)
Comment (Zombiepro89) [reply notamFetch.py:25]: Okay, I can do that. That's a pretty simple fix. Would the Notam object also utilize this new object?

File: notamFetch.py (hunk 1 above)
Comment (Zombiepro89) [reply notamFetch.py:116]: Yeah, I wasn't sure what to call this function. I can't think of a good name for it at the moment, do you have any suggestions?

File: notamFetch.py (hunk 4 above)
Comment (JamesJWilliams3000) [reply notamFetch.py:47]: I went looking around, and it looks like as of Python 3.4 we can do overloading [(heres the documentation)](https://peps.python.org/pep-0443/). We need to import `functools`, but its a part of the Python standard library and I didn't need to install anything to run the example code from the link above.

So we could use something like

```
@singledispatch
def get_notams(request_longitude_latitude) -> list:

@get_notams.register(str)
def _(airport_code) -> list:
```

File: notamFetch.py (hunk 1 above)
Comment (JamesJWilliams3000) [reply notamFetch.py:116]: I actually think swapping the names might work, as `get_notams` only returns part of our complete notam list

File: notamFetch.py (hunk 2 above)
Comment (JamesJWilliams3000) [reply notamFetch.py:25]: I don't think so, as the Notam object just hold the properties meant for sorting/viewing, and the `location` property of the notam comes from the issuing airport/location (so the lat/long we use for middle points won't be equivalent to the `location` property that the faa returns). I think the `request_longitude_latitude` object would be used for calculating middle points (so any middle point(s) we find are stored as a `request_longitude_latitude` object), and for `get_notams`.

File: notamFetch.py (hunk 3 above)
Comment (JamesJWilliams3000) [reply notamFetch.py:31]: I think it was more the entire `(not (airport_code is None))`; double negatives are difficult to read. 
According to [PEP 8:](https://peps.python.org/pep-0008/#programming-recommendations)

> For sequences, (strings, lists, tuples), use the fact that empty sequences are false:
> ```
> # Correct:
> if not seq:
> if seq:
> # Wrong:
> if len(seq):
> if not len(seq):
> ```

Since `airport_code` will be a string, using ```if not airport_code: ``` should be equivalent. 
IMO, I prefer ```if airport_code == "":``` , as it makes it obvious we are checking for an empty string.

**Comment by tanminivan** [comment]
Good comments so far. I'm purposely withholding my comments to give others on the team a chance to review first. 🙂

File: notamFetch.py (hunk 4 above)
Comment (aidanperkins) [reply notamFetch.py:47]: you can also do type enforcement for the different functions like this.
`def get_notams(request_longitude_latitude: tuple) -> list:`
`def _(airport_code: str) -> list: `

File: notamFetch.py [hunk 5]
```diff
@@ -3,60 +3,122 @@
...
+        
+        # If this request fails, prompt the user and return None
+        except Exception as e:
+            print("Invalid airport.", airport_code, "was not found.")
+            return None
+            
+        
+    # If the user did not enter an airport code or geolocator successfully outputted something,
+    # check to make sure the request is in the correct form
+    if(type(request_longitude_latitude) is tuple and len(request_longitude_latitude) == 2
```
Comment (aidanperkins) [inline notamFetch.py:44]: Since we raise an exception and return nothing when the request fails do we still need to check that the request_longitude_latitude is in the correct form?

File: notamFetch.py (hunk 1 above)
Comment (aidanperkins) [reply notamFetch.py:116]: Maybe change `get_notams` to `get_notam_at` and change `start_notam_request_path` to `get_notams` or `get_all_notams`.

File: notamFetch.py (hunk 2 above)
Comment (aidanperkins) [reply notamFetch.py:25]: For James's idea you could create a Point object from geopy with the coordinates without needing to make our own object class, or we could leave `request_longitude_latitude` as a float tuple and rename it `request_coords`.

Comment (JamesJWilliams3000) [reply notamFetch.py:25]: [Heres the geopy Point documentation](https://geopy.readthedocs.io/en/stable/index.html?highlight=point#geopy.location.Location.point). It does looks like a geopy Point object is what we need here.

**Review by MinhTeeFresh — COMMENTED** [review]
I would suggest maybe being a bit consistent about method order. From my understanding, get_notams_at utilizes 'send_api_request', which is below the method. However, the get_notams_at.register(str) utilizes get_notams_at, which is above the method. Not sure if there is a order requirement based on the functool library.

**Comment by Zombiepro89** [comment]
> I would suggest maybe being a bit consistent about method order. From my understanding, get_notams_at utilizes 'send_api_request', which is below the method. However, the get_notams_at.register(str) utilizes get_notams_at, which is above the method. Not sure if there is a order requirement based on the functool library.

Yeah, for functools library the generic function appear before the overloaded variants. But I think you're right that we should have some kind of consistent ordering. So I think i'll move send_api_request() above get_notams_at(), that way all functions appear before functions that they call. This ordering is similar to how C's scope works.

**Review by aidanperkins — APPROVED** [review]

**Review by Zenpock — APPROVED** [review]

File: notamFetch.py [hunk 6]
```diff
@@ -1,71 +1,132 @@
...
+        current_notam_properties = notam.get("properties").get("coreNOTAMData").get("notam")
         current_notam_text = current_notam_properties.get("text")
-        full_notam_list.append(current_notam_text)
+        notam_list.append(current_notam_text)
     
-    return full_notam_list
+    return notam_list
+
+
+    # TODO: Currently returns the union of the depature and arrival airport notams.
```
Comment (tanminivan) [inline notamFetch.py:107]: "TODO Currently returns the union of the departure and arrival airport notams"
Is this really a TODO or is it describing what is already done?

I understand comments like these help establish a direction/outline of what needs to happen but I think in practice they're problematic. It's too easy to leave them behind when code changes. It's also unclear when that comment should get removed, like in the case of KAN-39. Planning belongs in our planning tool, Jira. What if a comment here conflicts with what is in Jira?

I think it's fine to use this approach while you're developing—making comments that describe what some functions should do, describing in plain english or pseudocode. But it probably shouldn't make it to code review.

A (debatable) better strategy might be to stub out some of that desired functionality:

```python
def get_notams_along_path( start, end ):
    raise NotImplementedError( "Getting notams along path not yet implemented, see KAN-5" )

def get_all_notams( departure_airport, arrival_airport ):
   notams = []
   notams.append( get_notams_at( departure_airport ) )
   notams.append( get_notams_along_path( departure_airport, arrival_airport )
   notams.append( get_notams_at( arrival_airport )
   return sort_notams( notams )
```
This way, it's clear in the code what the desired form is. Downside is this will raise an exception when you try to run it. Alternatives might be for `get_notams_along_path()` to return an empty list, or a dummy list of notams: `["Dummy NOTAM 1 along path", "Dummy NOTAM 2 along path"]` if you want to have running code.

Alternatively, put the comments where that code would go so at least it's easier to clean up later:

```python
def get_all_notams( departure_airport, arrival_airport ):
   departure_airport_notams = get_notams_at( departure_airport )
   # TODO: Support getting notams along a path here (see KAN-5)
   arrival_airport_notams = get_notams_at( arrival_airport )
   # TODO: Support sorting notams here, after we've retrieved all necessary notams
   return departure_airport_notams + arrival_airport_notams
```

File: notamFetch.py [hunk 7]
```diff
@@ -1,71 +1,132 @@
...
 from dotenv import load_dotenv
+import Notam
+
 
 # load in env files
 load_dotenv()
 
 full_notam_list = []
 faa_api = "https://external-api.faa.gov/notamapi/v1/notams"
+geolocator = Nominatim(user_agent="notam_sort")
```
Comment (tanminivan) [inline notamFetch.py:16]: If this will function as a constant it should probably be stylized as such, like `GEOCODER`.

**Review by tanminivan — CHANGES_REQUESTED** [review]

**Comment by tanminivan** [comment]
Well, a push while I had some comments in review has messed up some of the line numbers and lost some of my comments. I'll try to rewrite them later today.

File: notamFetch.py [hunk 8]
```diff
@@ -1,71 +1,132 @@
...
+    """
+    From a given geopy Point object, send a request to the API.
+
+    request_longitude_latitude expects to be of a geopy Point object, which holds the longitude and latitude.
+    """
+    return send_api_request(request_longitude_latitude)
+
+
+@get_notams_at.register(str)
+def _(airport_code : str) -> list:
```
[the remark this answers is not in the document]
Comment (Zombiepro89) [reply notamFetch.py]: This is similar to what we had before, but I had changed it in response to previous comments. Would it be wise to revert this overloading that we're doing to keep in-line with python syntax?

[the remark this answers is not in the document]
Comment (tanminivan) [reply notamFetch.py]: Which is easier to read and understand?

What is the goal here? Is the goal to support fetching notams for both airport codes and coordinates? or is the goal to have function overloading?

File: notamFetch.py [hunk 9]
```diff
@@ -1,71 +1,132 @@
...
+    airport_code is expected to be a string."""
+
+    request_longitude_latitude = None
+
+    # Attempt to get the longitude and latitude of the airport.
+    try:
+        airport_info = geolocator.geocode(airport_code)
+        request_longitude_latitude = Point(airport_info.latitude, airport_info.longitude)
+    
+    # If this request fails, prompt the user and return None
```
[the remark this answers is not in the document]
Comment (Zombiepro89) [reply notamFetch.py]: I'm a little confused by both your comments here. Maybe I used the wrong term, but after this code the program prints and error, so it informs the user that their specified airport code was not found. 

As for your second, the code doesn't support it now, because we don't have any middle point computations, but I imagine if this function returns 'None' then we should signal that as something went wrong. If we get a valid input and no notams, it would probably return a blank array instead.

File: notamFetch.py [hunk 10]
```diff
@@ -1,71 +1,132 @@
...
-    api_result = requests.get(url=faa_api, params=faa_api_input , headers=credentials)
+    # Requesting the api for the notams
+    api_result = requests.get(url=faa_api, params=faa_api_input, headers=credentials)
     api_output = api_result.text
     api_status = api_result.status_code
 
+
+    # Code 200 indicates that the request occured successfully,
+    # if it's not 200 something went wrong.
+    # TODO: Create a more robust error system
```
[the remark this answers is not in the document]
Comment (Zombiepro89) [reply notamFetch.py:49]: KAN-39 didn't have a PR at the time I setup this PR, but i'll make sure to remove this comment when I update the code.

File: notamFetch.py [hunk 11]
```diff
@@ -1,71 +1,132 @@
...
-        current_notam = departure_notam_list[i]
-        # dig into the api response to get the important information
-        current_notam_properties = current_notam.get("properties").get("coreNOTAMData").get("notam")
+    # Get all of the notams and place them inside of notam list.
+    for notam in returned_notam_list:
+        
+        current_notam_properties = notam.get("properties").get("coreNOTAMData").get("notam")
         current_notam_text = current_notam_properties.get("text")
-        full_notam_list.append(current_notam_text)
+        notam_list.append(current_notam_text)
```
[the remark this answers is not in the document]
Comment (Zombiepro89) [reply notamFetch.py:70]: Yes, you're right. We should create a Notam object here instead of just returning the text.

**Comment by tanminivan** [comment]
I'm also curious when you plan to use airport codes vs coordinates. According to [this comment from Aiden](https://github.com/OU2024Capstone/notamSort/pull/7#discussion_r1493020318) it looks like you prefer to use coordinates over the airport codes so that you can get notams for the desired radius. Is that still true?

File: notamFetch.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (tanminivan) [reply notamFetch.py]: Probably the ambiguous term of "user" here. Will the user be the person at the command line terminal or the person interacting with the web application? 

> but I imagine if this function returns 'None' then we should signal that as something went wrong

Yes, precisely, and the way you signal that something went wrong is through raising exceptions :)

File: notamFetch.py [hunk 12]
```diff
@@ -1,71 +1,132 @@
...
+# this can be called anywhere doesnt have to be in the function call here
+# credentials retrieved from external '.env' file located in root dir
+credentials = {
+    "client_id": os.getenv('client_id'),
+    "client_secret": os.getenv('client_secret'),
+}
+
+
+@singledispatch
+def get_notams_at(request_longitude_latitude : Point) -> list:
```
[the remark this answers is not in the document]
Comment (tanminivan) [reply notamFetch.py]: This is a good question, and points out that everything is a tradeoff. Personally I prefer the simplicity of using primitives, but `Point`s _are_ easier to read but do expose you to some potential downsides.

[the remark this answers is not in the document]
Comment (tanminivan) [reply notamFetch.py]: I think there might be a good solution here by using a wrapper class.

File: notamFetch.py [hunk 13]
```diff
@@ -3,69 +3,145 @@
...
+
 
 # load in env files
 load_dotenv()
 
 full_notam_list = []
 faa_api = "https://external-api.faa.gov/notamapi/v1/notams"
+GEOLOCATOR = Nominatim(user_agent="notam_sort")
+DICT_KEY_LATITUDE = "latitude"
+DICT_KEY_LONGITUDE = "longitude"
```
[the remark this answers is not in the document]
Comment (Zombiepro89) [reply notamFetch.py:17]: It's just for convenience. If we ever need to change the keys for whatever reason, we just change these values instead of having to change every instance of this constant appearing.

File: notamFetch.py [hunk 14]
```diff
@@ -3,69 +3,145 @@
...
+
+        request_latitude_longitude = {DICT_KEY_LATITUDE: airport_info.latitude, DICT_KEY_LONGITUDE: airport_info.longitude}
+            
+
+    elif isinstance(location, dict):
+
+        if(DICT_KEY_LATITUDE in location.keys() and DICT_KEY_LONGITUDE in location.keys()):
+            
+            if(type(location[DICT_KEY_LATITUDE]) is float and type(location[DICT_KEY_LONGITUDE]) is float):
+                request_latitude_longitude = {DICT_KEY_LATITUDE: location[DICT_KEY_LATITUDE], DICT_KEY_LONGITUDE: location[DICT_KEY_LONGITUDE]}
```
[the remark this answers is not in the document]
Comment (Zombiepro89) [reply notamFetch.py:106]: I'm doing this just in case the 'location' dictionary has extra keys. It might be an unnecessary precaution though...? Though, we don't want to call the return statement here as the return statement at the end always occurs. So it would be a waste to add another return statement here.

File: notamFetch.py [hunk 15]
```diff
@@ -3,69 +3,145 @@
...
+            
+            else:
+                raise TypeError(f"Latitude and Longitude expected type float, but got {type(location[DICT_KEY_LATITUDE])} and {type(location[DICT_KEY_LONGITUDE])} instead.")
+
+        else:
+            raise KeyError(f"Keys are expected to be formatted as '{DICT_KEY_LATITUDE}' and '{DICT_KEY_LONGITUDE}'")
```
[the remark this answers is not in the document]
Comment (Zombiepro89) [reply notamFetch.py:112]: It prints "KeyError: Keys are expected to be formatted as 'latitude' and 'longitude'.". Aside from including the dictionary's name or saying "Input dictionary's keys ...", this can't really be more clear.

File: notamFetch.py [hunk 16]
```diff
@@ -3,69 +3,145 @@
...
-        full_notam_list.append(current_notam_text)
+        notam_list.append(current_notam_text)
     
-    return full_notam_list
+    print(f"Found {len(notam_list)} notams at {request_latitude_longitude}")
+
+    return notam_list
+
+
+def get_notams_at(location) -> list:
```
[the remark this answers is not in the document]
Comment (Zombiepro89) [reply notamFetch.py:74]: I was unaware of this, i'll amend this to help readability.

File: notamFetch.py [hunk 17]
```diff
@@ -3,69 +3,145 @@
...
+        
+        RuntimeError may occur if the inputted airport code is invalid.
+
+        TypeError may occur if the function recieves arguments that are not of type 'str' or 'dict'. 
+        May also occur if the dictionary keys 'latitude' or 'longitude' contain a value that is not of type 'float'
+        
+        KeyError may occur if the dictionary is missing either the 'latitude' or 'longitude' keys"""
+
+    if isinstance(location, str):
+        request_latitude_longitude = None
```
[the remark this answers is not in the document]
Comment (Zombiepro89) [reply notamFetch.py:89]: Ah, good catch. I missed this line of code. It should not be there, it's a vestige of older code.

File: notamFetch.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (aidanperkins) [reply notamFetch.py:112]: Maybe we should hint about the type associated to each key as well so the error has all the necessary information about the dictionary. 
Something like "KeyError: Dictionary keys are expected to be formatted as {'latitude' : float, 'longitude' : float}."

File: notamFetch.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (aidanperkins) [reply notamFetch.py:17]: Python standards were pretty unclear about if dictionary keys should be constants or not, but I think we could make the later code a little neater if we drop the `DICT_KEY_` portion and it would easier to type out.

File: notamFetch.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (aidanperkins) [reply notamFetch.py:106]: If we return here in the function the program wont call the return statement at the end, and as long as all the necessary keys are present with the correct types I don't really see a need to pull those into a new dictionary.

File: notamFetch.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (tanminivan) [reply notamFetch.py:67]: Probably out of scope for this ticket, better to get this closed and work on that in a future ticket. Probably a bit of work.

File: notamFetch.py [hunk 18]
```diff
@@ -3,69 +3,145 @@
...
+
+        # Attempt to get the longitude and latitude of the airport.
+        airport_info = GEOLOCATOR.geocode(location)
+
+        # If the geolocator outputs None, then something has gone wrong.
+        if(airport_info is None):
+            raise RuntimeError(f"Invalid airport. {location} was not found.")
```
Comment (tanminivan) [inline notamFetch.py:96]: Yes, nice

File: notamFetch.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (tanminivan) [reply notamFetch.py:112]: Yeah, that makes sense @aidanperkins

File: notamFetch.py [hunk 19]
```diff
@@ -3,69 +3,145 @@
...
+                request_latitude_longitude = {DICT_KEY_LATITUDE: location[DICT_KEY_LATITUDE], DICT_KEY_LONGITUDE: location[DICT_KEY_LONGITUDE]}
+            
+            else:
+                raise TypeError(f"Latitude and Longitude expected type float, but got {type(location[DICT_KEY_LATITUDE])} and {type(location[DICT_KEY_LONGITUDE])} instead.")
+
+        else:
+            raise KeyError(f"Keys are expected to be formatted as '{DICT_KEY_LATITUDE}' and '{DICT_KEY_LONGITUDE}'")
+
+    else:
+        raise TypeError(f"Unsupported input for location {location}, expected type str or dict but got {type(location)} instead.")
```
[the remark this answers is not in the document]
Comment (tanminivan) [reply notamFetch.py:115]: I could see both ways. Sometimes having the value makes it really clear where the lines got crossed. I'd probably put it in quotes though, so it's clear where the value ends and the log message begins.

**Review by tanminivan — APPROVED** [review]

**Comment by tanminivan** [comment]
Don't forget to squash and rebase.

**Comment by tanminivan** [comment]
SHIP IT!

**Review by aidanperkins — APPROVED** [review]

**Review by JamesJWilliams3000 — APPROVED** [review]
